modules/system/classes/asset/BundleManager.php

```diff
@@ -0,0 +1,253 @@
+                        'mix.js(\'assets/src/js/{{packageName}}.js\', \'assets/dist/js/{{packageName}}.js\').vue({ version: 3 });',
+                        $contents
+                    ),
+                    'js' => $this->getFixture('js/vue.js.fixture'),
+                    default => $contents
+                };
+            });
+        });
+    }
+
+    /**
+     * Loads registered asset bundles from modules and plugins
+     */
+    public function loadRegisteredBundles(): void
+    {
+        foreach ($this->callbacks as $callback) {
+            $callback($this);
+        }
+    }
+
+    /**
+     * Returns a list of the registered asset bundles.
+     */
+    public function listRegisteredBundles(): array
+    {
+        if (empty($this->registeredBundles)) {
+            $this->loadRegisteredBundles();
+        }
+
+        return $this->registeredBundles;
+    }
+
+    /**
+     * Get all bundles configured
+     */
+    public function getBundles(): array
+    {
+        return array_keys($this->listRegisteredBundles());
+    }
+
+    /**
+     * Get the packages for a bundle, with compiler specific packages
+     */
+    public function getBundlePackages(string $name, string $assetType): array
+    {
+        $config = $this->listRegisteredBundles()[$name] ?? [];
+
+        $packages = [];
+        foreach ($config as $key => $value) {
+            // Skip handlers
+            if (in_array($key, [static::HANDLER_SETUP, static::HANDLER_SCAFFOLD])) {
+                continue;
+            }
+
+            // Merge in any compiler specific packages for the current compiler
+            if (is_array($value)) {
+                if ($key === $assetType) {
+                    $packages = array_merge($packages, $value);
+                }
+                continue;
+            }
+
+            $packages[$key] = $value;
+        }
+
+        return $packages;
+    }
+
+    /**
+     * Registers a callback function that defines asset bundles. The callback function
+     * should register bundles by calling the manager's registerBundles() function.
+     * This instance is passed to the callback function as an argument. Usage:
+     *
+     *     BundleManager::registerCallback(function ($manager) {
+     *         $manager->registerAssetBundles([...]);
+     *     });
+     *
+     */
+    public function registerCallback(callable $callback): void
+    {
+        $this->callbacks[] = $callback;
+    }
+
+    /**
+     * Registers asset bundles.
+     */
+    public function registerBundles(array $definitions)
+    {
+        $this->registeredBundles = array_replace_recursive($this->registeredBundles, $definitions);
+    }
+
+    /**
+     * Registers a single asset bundle.
+     */
+    public function registerBundle(string $name, array $definition): void
+    {
+        $this->registerBundles([$name => $definition]);
+    }
```

Comment:
I'd personally prefer that `registerBundles` method does a loop call to the `registerBundle` method (not the other way round) and some documentation on the `registerBundle` method be added so that someone visually parsing the code (or using an IDE) will have some idea of what the methods are expecting (even if it's simple and provides a link to the docs site)

Reply:
I don't really want to document this too much, I'd prefer people don't use it unless they've got a good reason.

Reply:
Then perhaps put that as a recommendation in the documentation, but I find our code is littered with functionality that is hardly documented and it annoys me to no end, so I don't want to add to that.

Reply:
> I'd personally prefer that `registerBundles` method does a loop call to the `registerBundle` method (not the other way round) and some documentation on the `registerBundle` method be added so that someone visually parsing the code (or using an IDE) will have some idea of what the methods are expecting (even if it's simple and provides a link to the docs site)

Added here: https://github.com/wintercms/winter/pull/1164/commits/acc59b9b5ca7aa8514114d25c231f9929f92f821

Reply:
Thanks for sorting out the looping, @jaxwilko. However, the documentation is still a bit sparse.

Would you be able to at least summarise what is a "setup" handler and a "scaffold" handler, when they are run, and the contexts that they support?